Add TensorExpression support for std::complex<double> and ComplexDataVector

macedo22 opened the pull request:
## Proposed changes

This PR (1) refactors `TensorExpression` code and tests to be more easily extendable to other types and (2) adds support for using `std::complex<double>`, `Tensor<std::complex<double>>`, and `Tensor<ComplexDataVector>` terms in `TensorExpression`s.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

macedo22 commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,497 @@
+};
+template <typename T>
+struct upcast_if_derived_vector_type<
+    T, typename std::enable_if<is_vector<T>::value>::type> {
```
`git` warns to use `Requires` instead of `std::enable_if`, but I wasn't able to figure out how to get it to work in a template specialization. Let me know how I can use `Requires` here or suppress the warning!

wthrowe replied:
I would do this using a plain specialization (untested):
```c++
// Can wrap the bool in a bool_constant if you need to manipulate this in metaprogramming
template <typename T, bool IsVector = is_vector<T>::value>
struct upcast_if_derived_vector_type {
  using type = T;
};
template <typename T>
struct upcast_if_derived_vector_type<T, true> {
  //...
};
```

macedo22 replied:
Thanks! I'll try this

macedo22 replied:
I also added a definition for the `upcast_if_derived_vector_type<T, false>` specialization when it's not a vector just to make it really clear what happens when

wthrowe replied:
Fine with me.

wthrowe commented on src/DataStructures/Tensor/Expressions/Evaluate.hpp:
```diff
@@ -351,6 +367,13 @@ void evaluate_impl(
+    ASSERT(get_size((*lhs_tensor)[0]) > 0,
+           "Tensors with vector components must be sized before calling "
+           "tenex::evaluate<...>("
+           "\tgsl::not_null<Tensor<VectorType, ...>*>, number).");
```
Is the tab character intentional here?  Seems an odd place for one.

macedo22 replied:
I put it here in the hopes that it would make the function more readable in an error message, but looking at this I realize I forgot to add newlines in front of the line above it and in front of the tab, so this would look weird as is.

I'll add newlines unless you think I should scrap the newlines and tabs altogether

wthrowe replied:
That makes sense.  Tabs and newlines sound fine.

wthrowe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -79,6 +82,47 @@ struct upcast_if_derived_vector_type<
 };
 /// @}
 
+/// \brief Whether or not a given type is the complex-valued parter to another
```
partner

wthrowe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_EvaluateComplex.cpp:
```diff
@@ -0,0 +1,463 @@
+// or both
+//
+// \details
+// The tests in this file are designed to test that these funamentally work:
```
fundamentally

wthrowe commented:
Squash.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines which types are allowed, whether operations which certain types are
```
`which certain` -> `with certain`

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+///
+/// \tparam X the arithmetic data type
+template <typename X>
+using is_supported_number_datatype =
```
Would it be easier to make these `static constexpr bool is_supported_number_datatype_v = ...::value;`? Or add that? That way you don't need to do `::value` in a bunch of places.

Related question: `std::is_same` inherits off of `std::bool_constant` (ish) and so do you need to write it this way at all? Can you not just do `std::is_same<X, double>`?

macedo22 replied:
Since I have a boolean expression `std::is_same_v<...> or std::is_same_v<...>`, I think I think I can't get rid of `std::bool_constant` because I need to wrap the expression with it. However, I imagine I can inherit from `std::bool_constant`. Does the following work for you or did you have something else in mind that I'm not following?:
```
template <typename X>
struct is_supported_number_datatype
    : std::bool_constant<std::is_same_v<X, double> or
                         std::is_same_v<X, std::complex<double>>> {};
```
Alternatively, I could define a specialization for each and do this instead where we _can_ inherit off of `std:is_same`:
```
template <typename X>
struct is_supported_number_datatype;

template <typename X>
struct is_supported_number_datatype : std::false_type {};
template <>
struct is_supported_number_datatype : std::is_same<X, double>> {};
template <>
struct is_supported_number_datatype : std::is_same<X, std::is_complex<double>>> {};
```

Let me know if you have a preference or a different idea

nilsdeppe replied:
https://en.cppreference.com/w/cpp/types/disjunction

macedo22 replied:
Wow, never seen this but this is a really nice thing to know about

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+///
+/// \tparam T the given type
+template <typename T>
+using is_vector = std::bool_constant<std::is_base_of_v<MarkAsVectorImpl, T>>;
```
Isn't `is_derived_of_vector_impl<T>` good enough (defined in VectorImpl.hpp)? Same question about making it `is_vector_v` (or adding that)

macedo22 replied:
Definitely didn't see this defined there, so yes mine should be scrapped

macedo22 replied:
Wait sorry, mine differs in that it evaluates to `true` for both base `VectorImpl` and derived types, not just derived. So I think I will need to keep mine, if I understand `is_derived_of_vector` correctly

nilsdeppe replied:
In what case are you dealing with `VectorImpl` directly?

macedo22 replied:
Sorry, just removed my comment. Based on the alias' name, I had thought that `VectorImpl<double, DataVector>` would not be `true` because it isn't a _derived_ type of `VectorImpl`, but I just tested it and it's good. My `is_vector` is redundant, so I'll remove it.

macedo22 replied:
I'm dealing with `VectorImpl` directly in a few places where I wanted to extract the underlying value type and vector type, like here:

```
template <typename ValueType, typename VectorType>
struct lhs_datatype_is_assignable_to_rhs_datatype_impl<
    VectorImpl<std::complex<ValueType>, VectorType>, ValueType>
    : std::true_type {};
```

but I'm still reading through your comments, so maybe after I am done reading them, this won't be true anymore

macedo22 replied:
But in terms of when checking if something is a vector with `is_vector`, I don't think I'm actually using `VectorImpl`s directly, if that's what you're asking.

Regardless, my `is_vector` can be scrapped because it is equivalent to the existing`is_derived_of_vector_impl<T>`.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+  // if we have a derived VectorImpl, get base type, else T is a base VectorImpl
+  // and we use that
+  using type =
+      tmpl::conditional_t<std::is_base_of_v<MarkAsVectorImpl, upcasted_type>,
```
`is_derived_of_vector_impl_v<T>` or the `is_vector_v` you (will) define above?

macedo22 replied:
I think this needs to stay as is because the key part here is that this `type` should be the base `VectorImpl` type. Since something like `is_derived_of_vector_impl_v<VectorImpl<double, DataVector>>` will be `true`, this would incorrectly assign this `type` to be the base blaze type like `blaze::CustomVector<...>`, which is not what I want

Let me know if I am misunderstanding. I basically want this to be able to take a type and if it is a derived `VectorImpl`, I want the base `VectorImpl` type, otherwise do nothing to the type.

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+/// No template specialization was matched, so it's not a known pairing
+template <typename X1, typename X2>
+struct get_binop_datatype_impl {
+  using type = std::bool_constant<false>;
```
we also have a `NoSuchType` in `Utilities/NoSuchType.hpp` to signal that some sort of processing failed/is invalid :) brigand uses the same approach (we stole it from them, actually :P )

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -0,0 +1,387 @@
+/// \tparam X2 the data type of the other operand
+template <typename X1, typename X2>
+struct binop_datatypes_are_supported {
+  using type = std::bool_constant<not std::is_same_v<
```
why not inherit from `std::is_same<...>`? That'll add `::type` automatically and is normally how this sort of thing is handled.

Apply this comment to anywhere else in this file that you can, please :)

macedo22 replied:
After thinking about this more, I don't think I follow how this could inherit from `std::is_same` since I need to check if something is `not std::is_same`. The best I can think of is keeping the `std::bool_constant` and inheriting from that like:

```
template <typename X1, typename X2>
struct binop_datatypes_are_supported
    : std::bool_constant<not std::is_same_v<
          typename get_binop_datatype_impl<
              typename upcast_if_derived_vector_type<X1>::type,
              typename upcast_if_derived_vector_type<X2>::type>::type,
          NoSuchType>> {};
```

Did you have something different in mind that I'm not thinking of or is this fine?

nilsdeppe replied:
nah that's fine :)

nilsdeppe replied:
https://en.cppreference.com/w/cpp/types/negation

macedo22 replied:
Thanks, I haven't heard of this either, but I'll do the negation

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -82,6 +85,47 @@ struct upcast_if_derived_vector_type<T, true> {
+                                   typename DataVector::BaseType>
+    : std::true_type {};
+
+/// \brief Whether or not a given type is the complex-valued parter to another
```
`parter` -> `partner`

nilsdeppe commented on src/DataStructures/Tensor/Expressions/DataTypeSupport.hpp:
```diff
@@ -98,9 +142,12 @@ struct upcast_if_derived_vector_type<T, true> {
 template <typename LhsDataType, typename RhsDataType>
-struct lhs_datatype_is_assignable_to_rhs_datatype_impl : std::false_type {};
+struct lhs_datatype_is_assignable_to_rhs_datatype_impl
+    : std::bool_constant<(
```
once these things all inherit from `std::BLAH` you can remove the `bool_constant` stuff here too. That's why the inheritance approach is preferred, it cascades to all derived type traits

nilsdeppe commented on src/DataStructures/VectorImpl.hpp:
```diff
@@ -33,6 +33,46 @@
+/// specialization or add a new one.
+///
+/// \tparam LhsDataType the type being assigned
+/// \tparam RhsDataType the type to convert to `LhsDataType`
```
Can all this be checked with expression SFINAE? Like:
```cpp
template <typename T, typename S, std::void_t<>> is_assignable : std::false_type {};

template <typename T, typename S, std::void_t<decltype(std::declval<T>() = std::declval<S>())>> is_assignable :  std::true_type {};
```
and for that matter, https://en.cppreference.com/w/cpp/types/is_assignable might work even. The downside of `std::is_assignable` is that you need full type definitions at places where it's checked, but that should be fine. I'm not sure if it silently fails if the full type definition isn't available, but I assume so and is probably the correct behavior.

macedo22 replied:
I will look into this and use it if it works!

macedo22 replied:
I tried these but the issue I ran into is the opposite of what I would have expected - that I think more assignments would be permitted than we want. Apparently, you can assign a `DataVector` to a `ComplexDataVector`, but I imagine that we don't want that because when would we want that? And I figure in most cases if someone does that, it's an accident and the mistake would be hidden.

Because of this, I did not end up using either of the suggestions here in my fixups.

nilsdeppe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_DataTypeSupport.cpp:
```diff
@@ -115,7 +115,7 @@ SPECTRE_TEST_CASE("Unit.DataStructures.Tensor.Expression.DataTypeSupport",
   test_is_supported_number_datatype<double, true>();
```
Generalyquestion for this file: this is a very large number of similar checks. Is it possible to generalize them to use things like `tmpl::for_each`-type loops to do the checks? E.g. you could have a `tmpl::list<tmpl::list<std::complex<double>, std::true_type>, ...>`? Or, there might be an even easier way to generalize it, not sure :) But given the length of this file, it might be worth thing about :)

macedo22 replied:
That's a good idea. I will try to make it more concise and general!

macedo22 replied:
I was able to condense parts of the file a bit but not a ton

nilsdeppe reviewed (changes requested):
Overall looks great! Just some thoughts on type trait definitions, etc. I'd suggest reading through all the comments before starting changes, just to make sure some later comment doesn't cause design collisions :)

macedo22 commented:
@nilsdeppe First fixup addresses your review requests and the second fixup adds support for `std::complex<double> * Tensor<DataVector>()`. When I was working with Alex Carpenter on psi4, it was convenient to be able to do this operation for coding the product `iy` where `y` is a `Tensor<DataVector>`, but not sure if it's something we want to support or not

I grouped all of the fixups in this way for reviewing but when I eventually squash, I will break them up and squash the pieces into the original commits they should belong to.

nilsdeppe commented on tests/Unit/DataStructures/Tensor/Expressions/Test_EvaluateComplex.cpp:
```diff
@@ -386,60 +409,68 @@ void test_evaluate_large_expressions(
   MAKE_GENERATOR(generator);
 
-  const size_t vector_size = 3;
+  const size_t vector_size = 2;
```
Why the size change?

macedo22 replied:
I just figured there were a lot of FLOPs happening in this test and this just made it slightly less, but there is not reason it can't still be 3

nilsdeppe replied:
That's fine, was just curious.

nilsdeppe reviewed (commented):
LGTM, please rebase and squash :)

macedo22 committed:
Make TensorExpression tests more generalizable and maintainable

macedo22 committed:
Make TensorExpression source code more generalizable and maintainable

These changes make it so that when adding or changing type support within TensorExpressions, future edits will hopefully be easier. This commit moves data type configuration to its own file, centralizes the logic shared across expression types for determining which operations can and can't be done based on data types, and generalizes operations to accodomate future types more easily.

macedo22 committed:
Add TensorExpression support for Tensors with type ComplexDataVector

macedo22 committed:
Support std::complex<double> numbers, Tensors in TensorExpressions

This enables using std::complex<double> numeric terms and as the data type of Tensor terms.

nilsdeppe commented:
@wthrowe would you mind giving feedback on this since the last time you looked? Thank you!
